### DISCUSSION TITLE
Tach simulator feature?

### DISCUSSION THREAD

**Discussion by wyattwww** [Abasz/ESPRowingMonitor#5] [Ideas]
Hi!  I'm starting to implement FTMS on a fork so the ESPRowingMonitor can work for apps to appear as a rowing machine/ergometer.  Does the current code have the ability to simulate activity to make the protocol testing easier?  If not I can implement a simulation mode...

Also FYI I have tuned the KayakPro settings much better with a reasonable inertia setting.  Please let me know if these settings will work with your algorithms - I think you had mentioned that it might be off?  https://github.com/wyattwww/openrowingmonitor/commit/433eb6205712db126d7d538ecbc7fcc4d9d90e93#diff-1f39ce5f38b9c42fd8eb7f3357e9c79a8125022b878ca7fb5cd8d76eaee432e1

Thanks!
Wyatt

**Comment by Abasz** [comment]
Those settings should work. But you need to chose stroke detection algorithm. With so few points I would go with SLOPE. 

There is no auto adjusting dragfactor option. It is always on.

Sure it is capable of simulating :) it not possible to develop without:

You add the delta times to the `test.array.h` located in the root of the `src` folder and then uncomment `// simulateRotation();` in the `main.cpp` file. 

Within the simulateRotation() method you can set how many loops by changeing the int value at the end of the if statement:

`if (micros() - lastTestRevTime < test[i] || i2 == 3)`

**Reply by wyattwww** [reply]
Amazing, thanks for the pointers!  I’ll keep you posted.

**Reply by Abasz** [reply]
Just FYI:

It can do offline/local simulation as well. I.e. when you try to calibrate, similarly to ORM its possible to replay sessions.

It is under e2e tests. In VS Code select e2e test build, it will create a run_e2e_test.out file in the `build/e2e` directory (note I develop on windows so you may need to change `make` path in the tasks.json as on windows mingw is used for gcc complier). 

Then you can actually add a text file with the delta times as argument to this out file when you run it similarly like in ORM. But it is possible with the `simulate` argument to actually simulate delta times within a range. If you look into the the `main.cpp` in the `test` folder you can see what needs to be changed.
